cstrml/market/src/lib.rs

```diff
@@ -199,20 +221,97 @@ decl_module! {
+
+            // 2. Reject a pledge which is considered to be _dust_.
+            if value < T::Currency::minimum_balance() {
+                Err(Error::<T>::InsufficientValue)?
+            }
+            // 3. Ensure provider has enough currency.
+            ensure!(value <= T::Currency::free_balance(&who), Error::<T>::InsufficientCurrency);
+
+            let mut pledge_ledger = Self::pledge_ledgers(&who);
+            // 4. Increase total value
+            pledge_ledger.total += value;
+
+            // 5 Upsert pledge ledger
+
+            Self::upsert_pledge_ledger(&who, &pledge_ledger);
+
+            // 6. Emit success
+            Self::deposit_event(RawEvent::PledgeSuccess(who));
+
+            Ok(())
+        }
+
+
+        /// Descrease pledge amount of currency for market order.
+        #[weight = 1_000_000]
+        pub fn cut_pledge(origin, #[compact] value: BalanceOf<T>) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            // 1. Check if provider is registered.
+            ensure!(<Providers<T>>::contains_key(&who), Error::<T>::NotProvider);
+
+            // 2. Reject a pledge which is considered to be _dust_.
+            if value < T::Currency::minimum_balance() {
+                Err(Error::<T>::InsufficientValue)?
+            }
+
+            // 3. Ensure value is smaller than unused.
+            let mut pledge_ledger = Self::pledge_ledgers(&who);
+            ensure!(value <= pledge_ledger.total - pledge_ledger.used, Error::<T>::InsufficientPledge);
+
+            // 4. Decrease total value
+            pledge_ledger.total -= value;
+
+            // 5 Upsert pledge ledger
+
+            Self::upsert_pledge_ledger(&who, &pledge_ledger);
+
+            // 6. Emit success
+            Self::deposit_event(RawEvent::PledgeSuccess(who));
+
+            Ok(())
+        }
+
+
         /// Register to be a provider, you should provide your storage layer's address info
         #[weight = 1_000_000]
-        pub fn register(origin, address_info: AddressInfo) -> DispatchResult {
+        pub fn pledge(
+            origin,
+            #[compact] value: BalanceOf<T>,
+            address_info: AddressInfo
+        ) -> DispatchResult {
             let who = ensure_signed(origin)?;
 
             // 1. Make sure you have works
             ensure!(T::OrderInspector::check_works(&who, 0), Error::<T>::NoWorkload);
 
-            // 2. Upsert provision
+            // 2. Ensure provider has enough currency.
+            ensure!(value <= T::Currency::free_balance(&who), Error::<T>::InsufficientCurrency);
+
+            // 3. Upsert provision
             <Providers<T>>::mutate(&who, |maybe_provision| {
                 if let Some(provision) = maybe_provision {
                     // Change provider's address info
                     provision.address_info = address_info;
                 } else {
+                    // 4. Prepare new pledge ledger
```

Comment:
original `register` func supports register several times. I keep this ability here. Should we have an extr func to change address info?

Reply:
yes, recover the `register` function, treat it like `validate` in staking